### PR TITLE
[13.0][MIG] stock_picking_invoicing

### PR DESCRIPTION
**Pull request by Rad0van**
Because I needed this and also existing PR #786 is failing all the checks (it is far from anywhere being functional), seems to be stalled I decided to give it a try.

### CODE REVIEW

**Review by rvalyi — APPROVED** [review]
there are a few recent small fixes on the 12.0 branch but I accept this as a good start for 13.0 and also a helpful step for 14.0.

**Comment by Rad0van** [comment]
@rvalyi thanx. I can bring up the fixes from 12.0 later. However I found out that I didn't migrate the demo data. When trying that for some reason I cannot seem to find a way to fix it. Would you care to have a look?

File: stock_picking_invoicing/models/stock_move.py [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+        :return: float
+        """
+        product = self.mapped("product_id")
+        product.ensure_one()
+        if inv_type in ("in_invoice", "in_refund"):
+            result = product.price
+        else:
+            # If partner given, search price in its sale pricelist
+            if partner and partner.property_product_pricelist:
+                product = product.with_context(
```
Comment (Cedric-Pigeon) [inline stock_picking_invoicing/models/stock_move.py:56]: This piece of code won't work if the pricelist is configured to show the discount to the customer.

**Comment by mbcosta** [comment]
@Rad0van in the last months fixes and improvements was made in v12, as result of module implementation in Brazilian Localization, by this reason please consider include here the follow PRs:

* [FIX] Invoice price_unit https://github.com/OCA/account-invoicing/pull/859
* [IMP] Copy invoice state in backorders and extra stock move in pickings https://github.com/OCA/account-invoicing/pull/894
* [FIX] Pickings with same Partner to create Invoice but the Partner to Shipping is different should not be grouping https://github.com/OCA/account-invoicing/pull/867
* [IMP+FIX] Return Picking Invocing Process and mapping Invoice Type https://github.com/OCA/account-invoicing/pull/896
* [FIX] Mapping Invoice Type https://github.com/OCA/account-invoicing/pull/899

**Review by renatonlima — CHANGES_REQUESTED** [review]
Hi Rad0van,

Can you cherry-pick latest commits from 12.0? As @mbcosta commented, there are some important improvements and fixes there, I listed all the commits you need cherry-pick:

5209d6fe1ba778e722c8d110f0eced138e309f8b
f4d018131b5a02b29ff187bbba92d13c855f22f3
34f1aa842d6ba0831f82a3b302ae953b8e856ab9
9cffc46862b44bcc8a1b784785d85eb736b5b2a8
8b81aedff30628229b385b5c1f6c2c2be3531f11
e67d051fb04964a0c11a610b55020480a3ba44fc
a8373a7f495d7669bfdabe27ce20ad0da0e29c8c
b5960ef5ed7349fa3e0d90c6746fbdb44bb22282
a1eb1546f9d37e69f793e1aed8bf1c1442311668
afa166c53e4041182dd31b42ff1e4190ed0af832
913e64f97aba6b3b4a2b3e9c03bd2bb0f0684e2b
ea693a9bbae6e23e3bdf77686ec3f2fff81ef068
af13f9d6ca1cb24d587c5a80da8f194d53dd4d26
70057fba82ac66d0dce8da83b689db4b3c24b085
1289215e49a27823b6747bf6c7e6291caadf86d2
66b8c2eccc22f340d1b1419313fc54f8dd3343b5
7d7cedda99c33fc01a5c31bb0c6748deb3943c03
4644603d113ac182a2c4949e8d565e79bdc5e8e6
2baf273f698e0203490234c0f2d46c611c556d8c

**Comment by Rad0van** [comment]
> Hi Rad0van,
> 
> Can you cherry-pick latest commits from 12.0? As @mbcosta commented, there are some important improvements and fixes there, I listed all the commits you need cherry-pick:
> 
> [5209d6f](https://github.com/OCA/account-invoicing/commit/5209d6fe1ba778e722c8d110f0eced138e309f8b)
> [f4d0181](https://github.com/OCA/account-invoicing/commit/f4d018131b5a02b29ff187bbba92d13c855f22f3)
> [34f1aa8](https://github.com/OCA/account-invoicing/commit/34f1aa842d6ba0831f82a3b302ae953b8e856ab9)
> [9cffc46](https://github.com/OCA/account-invoicing/commit/9cffc46862b44bcc8a1b784785d85eb736b5b2a8)
> [8b81aed](https://github.com/OCA/account-invoicing/commit/8b81aedff30628229b385b5c1f6c2c2be3531f11)
> [e67d051](https://github.com/OCA/account-invoicing/commit/e67d051fb04964a0c11a610b55020480a3ba44fc)
> [a8373a7](https://github.com/OCA/account-invoicing/commit/a8373a7f495d7669bfdabe27ce20ad0da0e29c8c)
> [b5960ef](https://github.com/OCA/account-invoicing/commit/b5960ef5ed7349fa3e0d90c6746fbdb44bb22282)
> [a1eb154](https://github.com/OCA/account-invoicing/commit/a1eb1546f9d37e69f793e1aed8bf1c1442311668)
> [afa166c](https://github.com/OCA/account-invoicing/commit/afa166c53e4041182dd31b42ff1e4190ed0af832)
> [913e64f](https://github.com/OCA/account-invoicing/commit/913e64f97aba6b3b4a2b3e9c03bd2bb0f0684e2b)
> [ea693a9](https://github.com/OCA/account-invoicing/commit/ea693a9bbae6e23e3bdf77686ec3f2fff81ef068)
> [af13f9d](https://github.com/OCA/account-invoicing/commit/af13f9d6ca1cb24d587c5a80da8f194d53dd4d26)
> [70057fb](https://github.com/OCA/account-invoicing/commit/70057fba82ac66d0dce8da83b689db4b3c24b085)
> [1289215](https://github.com/OCA/account-invoicing/commit/1289215e49a27823b6747bf6c7e6291caadf86d2)
> [66b8c2e](https://github.com/OCA/account-invoicing/commit/66b8c2eccc22f340d1b1419313fc54f8dd3343b5)
> [7d7cedd](https://github.com/OCA/account-invoicing/commit/7d7cedda99c33fc01a5c31bb0c6748deb3943c03)
> [4644603](https://github.com/OCA/account-invoicing/commit/4644603d113ac182a2c4949e8d565e79bdc5e8e6)
> [2baf273](https://github.com/OCA/account-invoicing/commit/2baf273f698e0203490234c0f2d46c611c556d8c)

@renatonlima done. However this is quite blind-sided attempt - I cannot swear I understand all the changes. Will try to fix the issues tomorrow.

**Comment by Rad0van** [comment]
So I obviously need to fix in commit [05eb851] Now I am far from being expert on git so help me out here please. What is the way to change the contents of that commit? Would it be: `git rebase --interactive '05eb851^'` ?

**Comment by Rad0van** [comment]
@renatonlima , @mbcosta I have finished cherry-picking commits, adjusting things where necessary. Tests pass. So I guess this should be fine. Have to have a look at @Cedric-Pigeon comment and resolve that.

File: stock_picking_invoicing/models/stock_move.py (hunk 1 above)
Comment (Rad0van) [reply stock_picking_invoicing/models/stock_move.py:56]: I think I understand what you mean. Would using `def get_product_price(self, product, quantity, partner, date=False, uom_id=False)` of `product.pricelist` to get the price solve this?

Comment (Cedric-Pigeon) [reply stock_picking_invoicing/models/stock_move.py:56]: you should have a look here: https://github.com/odoo/odoo/blob/ae2d4b8060fe4c6c6ebf04739d33bb150b923224/addons/sale/models/sale.py#L1750

**Comment by sergioeix** [comment]
Hi i was trying this migration and comparing with 12v i am missing "create draft invoces" option under action menu on delivery orders. Is there another flow to do theese invoices from the delivery tree view?

**Comment by Rad0van** [comment]
> Hi i was trying this migration and comparing with 12v i am missing "create draft invoces" option under action menu on delivery orders. Is there another flow to do theese invoices from the delivery tree view?

You're probably right. It seems I failed tto properly migrate the action in stock_picking_invoicing/wizards/stock_invoice_onshipping_view.xml Will try to fix that.

File: stock_picking_invoicing/models/stock_move.py (hunk 1 above)
Comment (mbcosta) [reply stock_picking_invoicing/models/stock_move.py:56]: @Cedric-Pigeon @Rad0van the case of use for me is Brazilian Localization, in this case the creation of invoice from stock.picking means the company should created a Fiscal Document/account.invoice for any product outside movement included between the same company but in different address and another cases where are not an Purchase or Sale Order, so in those case the Price should be the Product Standard (  brazilian localization change for this case https://github.com/OCA/l10n-brazil/blob/12.0/l10n_br_stock_account/models/stock_move.py#L131 ). To use the creation of invoice from stock.picking integrated with Sale and Purchase and put back the possibilite to force the creation of invoice only from stock.picking( like the old behaivor in v8 ) was created two modules but at brazilian localization https://github.com/OCA/l10n-brazil/tree/12.0/l10n_br_sale_stock https://github.com/OCA/l10n-brazil/tree/12.0/l10n_br_purchase_stock ( we think in extract parts of the modules from there and put in this repo, but in v14) and in those modules the price_unit used are the value of price_unit at Purchase or Sale line( where the question about discount are already treat ).  It is important hear about the other cases of use and what the value of price_unit should expect here or even have different values depending of the kind of the movement .Do you think we should change the calcule ? Something like this:
```
if partner and partner.property_product_pricelist:
    product = product.with_context(
        lang=partner.lang,
        partner=partner.id,
        quantity=qty,
        pricelist=partner.property_product_pricelist.id,
        uom=fields.first(self).product_uom.id,
        date_order=fields.first(self).picking_id.scheduled_date,
    )
    #
    company = fields.first(self).env.user.company_id
    Fpos_sudo = fields.first(self).env['account.fiscal.position'].sudo()
    fpos_id = Fpos_sudo.with_context(
         force_company=company.id).get_fiscal_position(partner.id)
    taxes = Fpos_sudo.browse(fpos_id).map_tax(
         product.sudo().taxes_id.filtered(
             lambda x: x.company_id == company), product, partner)
    tax_obj = fields.first(self).env['account.tax']
    price_unit = tax_obj._fix_tax_included_price_company(
        product.price, product.taxes_id, taxes,
        fields.first(self).company_id)
    result = price_unit
```
Should we include field for discount, or another one, in stock.move ?

**Comment by sergioeix** [comment]
Hi @Rad0van how its going on? did you could check the action?
B/R

**Comment by mbcosta** [comment]
Hi @Rad0van , please consider include here the improvement made in PR https://github.com/OCA/account-invoicing/pull/969

**Review by rvalyi — APPROVED** [review]
I propose to merge this PR and leave the port of further 12.0 improvements like #969 for a next PR as the half full glass is certainly better than the empty glass and not merging is already causing new 14.0 PRs (https://github.com/OCA/account-invoicing/pull/1010) not being based on this PR which is much worse than waiting for perfection upfront.

**Comment by rvalyi** [comment]
superseded by https://github.com/OCA/account-invoicing/pull/1012

**Pull request closed without merging**